Simpler light syntax (#560)
decentraland/documentation: _posts/blockchain-integration/2022-02-17-estates-marketplace-integrations.md

---
date: 2022-02-17

> title: Decentraland's Estate Marketplaces Integration

Reviewer: I would prefer something more descriptive like
- `How to integrate Decentraland's Estate with your own Marketplace`
- `Integrating Decentraland's Estate in your Marketplace`

description: Important things to have in mind if you want to integrate the Decentraland's Estate in your marketplace
categories:
  - blockchain-integration
type: Document
---

> ## Decentraland's Estate Marketplace Integration

Reviewer: ```suggestion
```

Try to not repeat the title

The Decentraland's Estate is an ERC721 compliant NFT and runs in Ethereum Mainnet. Therefore a lot of third party marketplaces can trade them. In order to do so, these marketplace must follow
a certain rules in order to make keep the trading secure for the users.

> Each [Estate](https://docs.decentraland.org/decentraland/faq/#what-is-an-estate) is a group of [LANDs](https://docs.decentraland.org/decentraland/faq/#what-is-land). That group of LANDs creates a _`fingerprint`_ for the Estate. The _`fingerprint`_ is a _bytes_ hash. Everytime the Estate changes by adding or removing a LAND, its _`fingerprint`_ changes too.

Reviewer: ```suggestion
As you may know, each [Estate](https://docs.decentraland.org/decentraland/faq/#what-is-an-estate) is a group of [LANDs](https://docs.decentraland.org/decentraland/faq/#what-is-land) and it have a _bytes_ hash associated to the group that it call _`fingerprint`_. Everytime the Estate changes by adding or removing a LAND, its _`fingerprint`_ changes too.
```

> For marketplaces, specially the ones without an scrow system, it is 100% recomended to have a record of the Estate's _`fingerprint`_ when someone list it for sale or make an offer. That way, when the order/offer is successfully executed, the current owner can't change the estate by trying to frontrun the order execution.

Reviewer: ```suggestion
For marketplaces, specially the ones without a scrow system, it is 100% recomended to have a record of the Estate's _`fingerprint`_ when someone list it for sale or make an offer. That way, when the order/offer is successfully executed, the current owner can't change the estate by trying to frontrun the order execution.
```

It would be nice a link that explain what a `scrow system` is

> ### Example

Reviewer: ```suggestion
## Example
```

Now that you don't have the repeated title you can remove a `#` in every subsequent

> #### Not use the Estate fingerprint

Reviewer: ```suggestion
### Not use the Estate fingerprint
```

> - Bob has the Estate1 with LAND (1,1) and (1,2). Estate1 fingerprint: hash1

Reviewer: ```suggestion
- Bob has the `AwesomeEstate` with LAND (1,1) and (1,2) and the fingerprint  `HASH111`
```

> - Bob add the LAND (1,3) to Estate1. The Estate1 has the LANDs: (1,1), (1,2), and (1,3). Estate1 fingerprint: hash2 (Fingerprint changed)

Reviewer: ```suggestion
- Bob adds the LAND (1,3) to `AwesomeEstate`, now composed by the LANDs: (1,1), (1,2) and (1,3) and its fingerprint has changed to `HASH112`
```

> - Bob remove the LAND (1,1) from Estate1. The Estate1 has the LANDs: (1,2) and (1,3). Estate1 fingerprint: hash3 (Fingerprint changed)

Reviewer: ```suggestion
- Bob removes the LAND (1,1) from `AwesomeEstate`, now composed by the LANDs: (1,2) and (1,3) and its fingerprint has changed to `HASH113`
```

> - Bob put on sale the Estate1. The list is created onchain in the Ethereum mainnet with the Estate smart contract address, the Estate id, the price and the expiration date.

Reviewer: ```suggestion
- Bob puts `AwesomeEstate` up for sale. The list is created onchain in the Ethereum mainnet with the Estate smart contract address, the Estate id, the price and the expiration date.
```

> - Alice send a transaction to buy the Estate specifiyng the estate id and the price expected to pay.

Reviewer: ```suggestion
- Alice sends a transaction to buy the Estate specifying the Estate id and the price expected to pay.
```

> - Bob detects that someone is trying to buy his Estate1 and send a transaction with higher gas fees than Alice to remove the LANDs (1,2) and (1,3) from Estate1.

Reviewer: ```suggestion
- Bob detects that someone is trying to buy his `AwesomeEstate` and sends a transaction with higher gas fees than Alice to remove the LANDs (1,2) and (1,3) from `AwesomeEstate`.
```

> - Bob's transactions is mined first. Estate1 has 0 LANDs. Estate1 fingerprint: hash4 (Fingerprint changed)

Reviewer: ```suggestion
- Bob's transactions is mined first. `AwesomeEstate` has 0 LANDs and its fingerprint has changed to `HASH114`
```

> - Alice's transaction is mined later. Alice bought the Estate1 with 0 LANDs on it. It means that Alice got front-runned (and steal/scammed) by Bob.

Reviewer: ```suggestion
- Alice's transaction is mined later. She bought the `AwesomeEstate` with 0 LANDs on it. It means that Alice got front-runned (and steal/scammed) by Bob.
```

> #### Use the Estate fingerprint

Reviewer: ```suggestion
### Use the Estate fingerprint
```

> - Bob has the Estate1 with LAND (1,1) and (1,2). Estate1 fingerprint: hash1

Reviewer: ```suggestion
- Bob has the `AmazingEstate` with LAND (1,1) and (1,2) and the fingerprint `HASH111`
```

> - Bob add the LAND (1,3) to Estate1. The Estate1 has the LANDs: (1,1), (1,2), and (1,3). Estate1 fingerprint: hash2 (Fingerprint changed)

Reviewer: ```suggestion
- Bob adds the LAND (1,3) to `AmazingEstate`. The Estate1 has the LANDs: (1,1), (1,2), and (1,3) and its fingerprint has changed to `HASH112`
```

> - Bob remove the LAND (1,1) from Estate1. The Estate1 has the LANDs: (1,2) and (1,3). Estate1 fingerprint: hash3 (Fingerprint changed)

Reviewer: ```suggestion
- Bob removes the LAND (1,1) from `AmazingEstate`, now composed by the LANDs: (1,2) and (1,3) and its fingerprint has changed to `HASH113`
```

> - Bob put on sale the Estate1. The list is created onchain in the Ethereum mainnet with the Estate smart contract address, the Estate id, the price, the expiration date.

Reviewer: ```suggestion
- Bob puts `AmazingEstate` up for sale. The list is created onchain in the Ethereum mainnet with the Estate smart contract address, the Estate id, the price, the expiration date.
```

> - Alice send a transaction to buy the Estate specifiyng the estate id, the price expected to pay, and **the fingerprint she saw (hash3)**.

Reviewer: ```suggestion
- Alice sends a transaction to buy the `AmazingEstate` specifying the Estate id, the price expected to pay, and **the fingerprint she saw (`HASH113`)**.
```

> - Bob detects that someone is trying to buy his Estate1 and send a transaction with higher gas fees than Alice to remove the LANDs (1,2) and (1,3) from Estate1.

Reviewer: ```suggestion
- Bob detects that someone is trying to buy his `AmazingEstate` and sends a transaction with higher gas fees than Alice to remove the LANDs (1,2) and (1,3) from `AmazingEstate`.
```

> - Bob's transactions is mined first. Estate1 has 0 LANDs. Estate1 fingerprint: hash4 (Fingerprint changed)

Reviewer: ```suggestion
- Bob's transactions is mined first, now `AmazingEstate` has 0 LANDs and its fingerprint has changed to `HASH114`
```

> - Alice's transaction is reverted because the smart contract checked that the fingerprint in the param that Alice sent didn't match with the current Estate1 fingerprint (hash3 != hash4). This check prevented Alice to buy a non desired Estate.

Reviewer: ```suggestion
- Alice's transaction is reverted because the smart contract checked that the fingerprint in the param that Alice sent didn't match with the current `AmazingEstate` fingerprint (`HASH113` != `HASH114`). This check prevented Alice to buy a non desired Estate.
```

> ### Estate Smart Contract Interface

Reviewer: ```suggestion
## Estate Smart Contract Interface
```

> The Estate's smart contract is compliant with a [fingerprint interface](https://github.com/decentraland/land/blob/master/contracts/estate/EstateStorage.sol#L19). In order to check if an order/offer for an estate is still valid or not you can call the _`verifyFingerprint(uint256 estateId, bytes fingerprint)`_ function implemented in the Estate smart contract. You can check a working production example [here](https://github.com/decentraland/marketplace-contracts/blob/master/contracts/marketplace/MarketplaceV2.sol#L382)

Reviewer: ```suggestion
The Estate's smart contract is compliant with a [fingerprint interface](https://github.com/decentraland/land/blob/master/contracts/estate/EstateStorage.sol#L19) and in order to check if an order/offer for an estate is still valid or not you can call the _`verifyFingerprint(uint256 estateId, bytes fingerprint)`_ function implemented in the Estate smart contract. You can check a working production example [here](https://github.com/decentraland/marketplace-contracts/blob/master/contracts/marketplace/MarketplaceV2.sol#L382)
```
